File: HOT100/HOT100p1/12.栈/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        pair = {'(' : ')', '[' : ']', '{' : '}'}
        for ch in s:
            if ch in pair:
                stk.append(ch)
            elif ch in pair.values():
                if not stk or pair[stk[-1]] != ch: return False
                stk.pop()
            else: return False
        return not stk
# @lc code=end

File: HOT100/HOT100p1/12.栈/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */
#include "tools.h"

// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> hash;
        stack<char> stk;
        hash['('] = ')', hash['['] = ']', hash['{'] = '}';
        for (auto ch : s)
        {
            if (hash.find(ch) != hash.end()) stk.push(ch);
            else if (ch == ')' || ch == ']' || ch == '}')
            {
                if (! stk.size() || hash[stk.top()] != ch) return 0;
                stk.pop();
            }
            else return 0;
        }
        return stk.empty();
    }
};
// @lc code=end

